scripts.py

```diff
@@ -707,6 +709,256 @@ def do_run(self):
+                          if int(migration[:8]) >= int(last_run)]
+
+        # Multiple migrations run on the same day have an additional digit
+        # after the date and a dash, eg:
+        #
+        #     20150826-1-change_target_age_from_int_to_range.sql
+        #
+        # When that migration is run, the number will be saved to the
+        # 'counter' column of Timestamp, so we have to account for that.
+        start_found = False
+        later_found = False
+        index = 0
+        while not start_found and not later_found and index < len(new_migrations):
+            start_found, later_found = self._is_matching_migration(
+                new_migrations[index], timestamp
+            )
+            index += 1
+
+        if later_found:
+            index -= 1
+        new_migrations = new_migrations[index:]
+        return new_migrations
+
+    def _is_matching_migration(self, migration_file, timestamp):
+        """Determine whether a given migration filename matches a given
+        timestamp or is after it.
+        """
+        is_match = False
+        is_after_timestamp = False
+
+        timestamp_str = timestamp.timestamp.strftime('%Y%m%d')
+        counter = timestamp.counter
+
+        if migration_file[:8]>=timestamp_str:
+            if migration_file[:8]>timestamp_str:
+                is_after_timestamp = True
+            elif counter:
+                count = self.MIGRATION_WITH_COUNTER.search(migration_file)
+                if count:
+                    migration_num = int(count.groups()[0])
+                    if migration_num==counter:
+                        is_match = True
+                    if migration_num > counter:
+                        is_after_timestamp = True
+            else:
+                is_match = True
+        return is_match, is_after_timestamp
+
+    def run_migrations(self, migrations, migrations_by_dir, timestamp):
+        """Run each migration, first by timestamp and then by directory
+        priority.
+        """
+        previous = None
+
+        migrations = self._sort_migrations_with_counter(migrations)
+        for migration_file in migrations:
+            for d in self.directories_by_priority:
+                if migration_file in migrations_by_dir[d]:
+                    full_migration_path = os.path.join(d, migration_file)
+                    self._run_migration(full_migration_path, timestamp)
+                    self._db.commit()
+                    previous = migration_file
+
+    def _sort_migrations_with_counter(self, migrations):
+        """Ensures that migrations with a counter digit are sorted after
+        migrations without one.
+        """
+        WITHOUT_COUNTER = "-(\D+).(py|sql)"
+        migrations = sorted(migrations)
+
+        # The built-in sorted() method places migrations that start with
+        # just a date (e.g. '20160825-do-this.sql') AFTER migrations
+        # with a counter digit (e.g. '20160825-1-and-that.py').
+        #
```

Comment:
I think you're missing an explanatory step before the 'therefore'. I think this is it:

If we run migrations in this order, `Timestamp.counter` will be unusable. It won't be clear whether `Timestamp.counter` is None because all the migration scripts ran, or because none of them ran.

Is that right?